Sources/SwiftSyntaxBuilder/ResultBuilders.swift.gyb

```diff
@@ -85,6 +85,11 @@ public struct ${type.result_builder()} {
   public static func buildFinalResult(_ component: Component) -> FinalResult {
 %   if element_type.is_token():
     return .init(component)
+%   elif element_type.has_with_trailing_comma_trait():
+    let lastIndex = component.count - 1
+    return .init(component.enumerated().map({ index, element in
+      element.create${element_type.buildable_base_name()}().withTrailingComma(index < lastIndex)
```

Comment:
I think I mentioned this before but unconditionally calling `withTrailingComma` here effectively prevents you from adding a trailing comma using result builders (because the result builder will remove the comma on the last element).

I think we don’t have this problem if this would be something like

```suggestion
      var buildable = element.create${element_type.buildable_base_name()}()
      if index < lastIndex {
        buildable = buildable.withTrailingComma(index < lastIndex)
      }
      return buildable
```

I think it would also be worth adding a test case that explicitly adds a trailing comma to the last element and assert that the trailing comma is retained through the result builder.

Reply:
I thought [you've agreed](https://github.com/apple/swift-syntax/pull/454#discussion_r895417798) that it's fine if trailing commas are preserved in the Element List types initializers, but always overridden in the Result Builders (in both ways added and removed).

Reply:
Sorry, I should have maybe been clearer (or I’m misunderstanding you). What I mean is that I think you should be able to generate e.g. an array literal with trailing comma like `[1,2,]` using the result builder syntax using something like the following (pseudo-code written in GitHub, probably doesn’t compile)
```swift
ArrayExpr {
  ArrayElement(IntegerLiteralExpr(1))
  ArrayElement(IntegerLiteralExpr(2), trailingComma: .comma)
}
```

So basically my thinking is that the result builder should only add commas but never remove any. My thinking is that some people prefer to have trailing commas in their array/dictionary literals and I don’t think we should deprive them of that ability.

Reply:
I don't see much value in it because Result Builders don't allow explicit trivias on non-last trailing commas anyway. I think list type initializer should be used in the scenario you described. But since you insist, I've updated the PR with that and I've added a [test](https://github.com/apple/swift-syntax/pull/454/files#diff-168af55d0fe26fcb8e58da2a1e597fa45b6d9cbc7832c1a1d5a4273d50b7737eR31) for it as well.

Reply:
I’m totally open to debate here. I would have just been surprised if the example I put above dropped the trailing comma. Do you disagree? Or do you have an example where the behavior you implemented before would be less surprising than what I’m suggesting?

Reply:
Trailing comma in a single line array literal (like `[1, 2, 3, ]`) doesn't make much sense to me, however I see the point in case of multiline:
```
[
  1,
  2,
  3,
]
```
However current implementation won't allow you to generate code like the second example with result builder, because it'll overwrite trailing commas removing explicit new line trivias.
It might be surprising, but I'd say that it works as designed: you either use convenience of result builder with no way to customize the output, or you use more verbose Element List initializer and has full control on the output.
Finally, I'd like to merge this PR ASAP and then the logic could be improved as a separate change

Reply:
OK, now I understand what you meant by “don't allow explicit trivias on non-last trailing commas”. With that in mind, I don’t have strong opinions in either direction anymore.

I’ve merged the PR as-is for now. If you want, I won’t object to reverting the behavior back to what you originally implemented. 😉